File: lib/get_playlist.dart
import 'dart:convert';
import 'dart:io';

import 'package:json/json.dart' show JsonDecodable;
import 'package:spot/config.dart' show Config;
import 'package:http/http.dart' as http;
import 'package:spot/item.dart' show Item;
import 'package:spot/track.dart' show Track;

@JsonDecodable()
class AccessTokenResponse {
  String access_token = '';
  String token_type = '';
  int expires_in = 0;
}

@JsonDecodable()
class ResponseData {
  List<Item> items = List.empty();
}

void main() async {
  final config = Config.parse('config.json');
  final items = await getPlaylistTracks(config, 3);
  final tracks = Track.fromItems(items);
  tracks.sort((a, b) => a.addedAt.compareTo(b.addedAt));
  final encoder = JsonEncoder.withIndent('    ');
  final json = encoder.convert(tracks);
  File('tracks.json').writeAsString(json);
}

Future<List<Item>> getPlaylistTracks(Config config, int tries) async {
  if (tries <= 0) {
    throw Exception('getPlaylistTracks out of tries');
  }

  final token = await getAccessToken(config, 3);
  final params = {
    'limit': '50',
    'fields':
        'items(added_at,track(id,name,track_number,external_urls(spotify),album(total_tracks,name,release_date,images(url)),artists(name)))'
  };

  final playlistUrl = '${config.apiUrl}/playlists/${config.playlistId}/tracks';
  final url = Uri.parse(playlistUrl).replace(queryParameters: params);
  final headers = {HttpHeaders.authorizationHeader: 'Bearer $token'};

  final res = await http.get(url, headers: headers);

  if (res.statusCode == 401) {
    await clearAccessToken();
    return getPlaylistTracks(config, tries - 1);
  }

  if (res.statusCode != 200) {
    throw Exception(res.body);
  }

  final json = jsonDecode(res.body);
  final data = ResponseData.fromJson(json);

  return data.items;
}

Future<String> getAccessToken(Config config, int tries) async {
  if (tries <= 0) {
    throw Exception('getAccessToken out of tries!');
  }

  final file = File('.access_token');
  var token = '';

  if (await file.exists()) {
    token = await file.readAsString();
  }

  if (token == '') {
    await fetchAccessToken(config);
    return getAccessToken(config, tries - 1);
  }

  return token;
}

Future<void> fetchAccessToken(Config config) async {
  final data = {
    'grant_type': 'client_credentials',
    'client_id': config.clientId,
    'client_secret': config.clientSecret
  };

  final url = Uri.https('accounts.spotify.com', '/api/token');
  final res = await http.post(url, body: data);
  final json = jsonDecode(res.body);
  final tokenData = AccessTokenResponse.fromJson(json);

  await File('.access_token').writeAsString(tokenData.access_token);
}

Future<void> clearAccessToken() async {
  File('.access_token').writeAsStringSync('');
}

File: src/get_playlist.rs
use crate::{item::Item, track::Track, utils::Result};
use reqwest::blocking;
use serde::Deserialize;
use serde_json::json;
use std::{cmp::Ord, env, fs};

const ACCESS_TOKEN_PATH: &str = ".access_token";

pub fn call() -> Result<()> {
    let config = Config::from_env()?;
    let items = get_playlist_tracks(&config, 3)?;

    let mut tracks = Track::from_items(items);
    tracks.sort_by(|a, b| Ord::cmp(&a.added_at, &b.added_at));

    let json = serde_json::to_string_pretty(&tracks)?;
    fs::write("tracks.json", json)?;

    Ok(())
}

#[derive(Debug)]
struct Config {
    playlist_id: String,
    api_url: String,
    client_id: String,
    client_secret: String,
}

#[derive(Debug, Deserialize)]
struct ResponseData {
    items: Vec<Item>,
}

fn get_playlist_tracks(config: &Config, tries: u32) -> Result<Vec<Item>> {
    if tries == 0 {
        return Err("get_playlist_tracks out of tries!".into());
    }

    let params = json!({
        "limit": 50,
        "fields": "items(added_at,track(id,name,track_number,external_urls(spotify),album(total_tracks,name,release_date,images(url)),artists(name)))"
    });

    let token = get_access_token(config)?;

    let url = format!("{}/playlists/{}/tracks", config.api_url, config.playlist_id);
    let req = blocking::Client::new()
        .get(url)
        .query(&params)
        .header("Authorization", format!("Bearer {}", token));

    let res = req.send()?;

    if res.status() == 401 {
        clear_access_token()?;
        return get_playlist_tracks(config, tries - 1);
    }

    if res.status() != 200 {
        let text = res.text()?;
        return Err(text.into());
    }

    let data: ResponseData = res.json()?;

    Ok(data.items)
}

fn get_access_token(config: &Config) -> Result<String> {
    let mut token = fs::read_to_string(ACCESS_TOKEN_PATH).unwrap_or("".into());

    if token.is_empty() {
        token = fetch_access_token(config)?;
        fs::write(ACCESS_TOKEN_PATH, &token)?;
    }

    Ok(token)
}

fn fetch_access_token(config: &Config) -> Result<String> {
    let data = json!({
        "grant_type": "client_credentials",
        "client_id": &config.client_id,
        "client_secret": &config.client_secret,
    });

    let req = blocking::Client::new()
        .post("https://accounts.spotify.com/api/token")
        .form(&data);

    let res = req.send()?;
    let json: AccessTokenResponse = res.json()?;

    Ok(json.access_token)
}

fn clear_access_token() -> Result<()> {
    fs::write(ACCESS_TOKEN_PATH, "").map_err(Into::into)
}

#[allow(dead_code)]
#[derive(Debug, Deserialize)]
struct AccessTokenResponse {
    pub access_token: String,
    pub token_type: String,
    pub expires_in: u64,
}

impl Config {
    fn from_env() -> Result<Self> {
        dotenvy::dotenv()?;

        let playlist_id = get_var("PLAYLIST_ID")?;
        let api_url = get_var_with_default("API_URL", "https://api.spotify.com/v1");
        let client_id = get_var("CLIENT_ID")?;
        let client_secret = get_var("CLIENT_SECRET")?;

        Ok(Self {
            playlist_id,
            api_url,
            client_id,
            client_secret,
        })
    }
}

fn get_var(name: &str) -> Result<String> {
    env::var(name).map_err(|_| format!("{name} required").into())
}

fn get_var_with_default(name: &str, default: &str) -> String {
    env::var(name).unwrap_or(default.to_owned())
}
